HSEARCH-3612 Normalize the terms to match in wildcard/prefix predicates

fax4ever opened the pull request:
https://hibernate.atlassian.net/browse/HSEARCH-3612

Tested locally for all Elasticsearch supported profiles: the behavior seems consistent among all of them.

yrodiere commented on backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/predicate/impl/LuceneTextFieldPredicateBuilderFactory.java:
```diff
@@ -61,7 +61,7 @@ public LuceneTextFieldPredicateBuilderFactory( boolean searchable,
 			String absoluteFieldPath) {
 		checkSearchable( absoluteFieldPath );
-		return new LuceneTextWildcardPredicateBuilder( absoluteFieldPath );
+		return new LuceneTextWildcardPredicateBuilder( absoluteFieldPath, analyzerOrNormalizer );
```
There's also a "prefix" syntax in the simple query string predicate. Could you please implement the fix and test it for that predicate, too?

fax4ever replied:
Sorry I missed completely the `prefix` part. Adding now...

yrodiere commented on documentation/src/main/asciidoc/search-dsl.asciidoc:
```diff
@@ -462,30 +462,24 @@ include::{sourcedir}/org/hibernate/search/documentation/searchdsl/predicate/Pred
-when targeting a field that applies a lowercase filter when indexing.
-`cat*` may match, since it's already lowercased.
+For example, a pattern such as `Cat*` could match `cat`
+when targeting a field having a normalizer applies a lowercase filter when indexing.
```
`normalizer applies` => `normalizer that applies`

yrodiere commented on documentation/src/main/asciidoc/search-dsl.asciidoc:
```diff
@@ -462,30 +462,24 @@ include::{sourcedir}/org/hibernate/search/documentation/searchdsl/predicate/Pred
+
+If an analyzer has been defined on the field,
+using Elasticsearch backend the patterns won't be analyzed nor normalized,
+using Lucene backend the patters will be normalized, but not tokenized.
```
`patters` => `patterns`

yrodiere replied:
I would rephrase:

```
If an analyzer has been defined on the field:

* when using the Elasticsearch backend, the patterns won't be analyzed nor normalized,
and will be expected to match a *single* indexed token, not a sequence of tokens.
* when using the Lucene backend the patterns will be normalized, but not tokenized:
the pattern will still be expected to match a *single* indexed token, not a sequence of tokens.
```

fax4ever commented:
@yrodiere According to the last test I added it seems that the terms are already normalized, even without the `analyze_wildcard` flag enabled. Addressed the comment so far

yrodiere commented:
@fax4ever To be honest, I didn't think prefix queries would perform tokenization at all. I expected them to only work with a single term, just like the wildcard queries. I was surprised that your tests passed...

fax4ever commented:
Indeed it doesn't work :) 

So let's just stick tot:
> Test normalization of prefix within simple query terms 

Is there anything else we can improve here?
Thanks

yrodiere commented:
> Indeed it doesn't work :)

I don't get it, what doesn't work?

I was talking about prefix queries using multiple terms, which I didn't think would work, e.g. : `"a few words"*`. I only expected single-term prefix queries to work, e.g. `word*`.

fax4ever commented:
Me too. I had tried a tess, such as: 
> "prefix1 prefix2 prefix3"*.

And it didn't work. For that reason I settled for normalization only.

yrodiere commented:
But your test contains this:

```
		assertThat( createQuery.apply( "\"" + PREFIX_1_DIFFERENT_CASE + "\"*" ) )
				.hasDocRefHitsAnyOrder( INDEX_NAME, DOCUMENT_1, DOCUMENT_2, DOCUMENT_3, DOCUMENT_4 ); 
```

The query string evaluates to `"SmILE on oUr LOves, and while thou dRAWest thE"*`, and the targeted field is analyzed, so the field itself is tokenized.

If the prefix query didn't tokenize the input, it would try to find a term starting with `SmILE on oUr LOves, and while thou dRAWest thE`, and it would fail, because the only terms in the index are `smile`, `on`, `our`, `loves`, etc., not the complete sentence.

So it *does* tokenize input. Which is nice, I suppose?

I launched a full build, let's see how it goes, but if everything passes, +1 to merge. Thanks!

fax4ever commented:
@yrodiere yes It is. Thank you. 
Good idea to lunch a full build to test the different version of Elasticsearch.

fax4ever commented:
Merged thanks!
